### ISSUE TITLE
Web Frameworks Library for MRPT (Rachit Tibrewal)

### ISSUE THREAD

**Issue by jolting** [MRPT/GSoC2018-discussions#6]
## Initial Description

Robots are slowly becoming a part of the internet of things, and easy control and universal access will be a step in future of robotics. MRPT’s Web Framework Library should address this problem.A lightweight publisher/subscriber mechanism needs to be created for MRPT. The C++ server library and the javascript library will then leverage this to create further applications. There will be two libraries, a C++ server library for RPC on the robot. Another will be a javascript library which provides modules for such procedure calls.This task involve serialization of existing mrpt objects for JSON transfer, RPC protocol, websocket setup, 3D rendering of objects in js, reusable js components for sending and receiving data from user created app where the user is largely benefitted from easy to use library methods and objects.

---

* Student: @rachit173
* Main Mentor: @jolting
* Backup Mentor: @jlblancoc

**Comment by rachittibrewal** [comment]
Hello @jolting, @jlblancoc Thanks for giving me this opportunity. I am very excited about the project and working with mrpt community. I would like to contribute to some issues and read codebase till May 5, as my exams will continue till that date, after that I will start working towards the first phase.

**Comment by jolting** [comment]
Hi @rachit173,
At this point I would just like to confirm that you're able to build MRPT. If you're using Linux I would recommend Ubuntu 18.04 since the compiler is already up to date. On Windows of course you'll need MSVC 2017. I don't personally develop on OS X, so I'm very little help there.

**Comment by tibrewalrachit** [comment]
I had built it about 20 days ago on 16.04, I will install 18.04 and build
MRPT.

**Comment by rachittibrewal** [comment]
Sorry for the confusion, i am creating by using my alternate GitHub account y73isc00l.I will continue to use rachit173. I am installing 18.04 now.

**Comment by rachittibrewal** [comment]
@jolting I have built mrpt-2.0.0 from source on Ubuntu 18.04. Building the apps was giving error so I turned it off for build. Also I tried to install ROS(kinetic-kame) on 18.04 but there seems to be no version available, is there a workaround?

**Comment by jolting** [comment]
ROS melodic is not ready yet. You can install a few packages from the archive.

If you need to run ROS kinetic you can use a docker container. 
http://wiki.ros.org/docker/Tutorials/GUI

Do you need that to compare against ros3djs?

**Comment by rachittibrewal** [comment]
Yes, since 3d visualisation is going to be similar to ros3djs,it would be better to have it in the same os. Also I was thinking about rosmrpt.

On Apr 27, 2018 07:01, Hunter Laux <notifications@github.com> wrote:

ROS melodic is not ready yet. You can install a few packages from the archive.

If you need to run ROS kinetic you can use a docker container. Melodic
http://wiki.ros.org/docker/Tutorials/GUI

Do you need that to compare against ros3djs?

**Comment by jolting** [comment]
Sure. MRPT does have some ROS bindings.

https://github.com/mrpt-ros-pkg/mrpt_navigation

ROS has a lot of packages and a large community of users. It's nice to be able to take advantage of all of that. I use ROS all the time. I encourage you to learn all you can from what the RobotWebTools guys have done with rosbridge_server and ros3djs.

However, we do want you to implement an MRPT specific version that is not coupled to ROS. Here are a few reasons:

1. I think potentially a non-ROS solution can be much more cohesive and usable with MRPT.
2. We want it to be much more portable than ROS.
3. PUB/SUB and RPC should be optimized for web. For example, rosbridge_server naively converts everything to JSON. This might not always be the right thing to do. You can send binary blobs over a websocket.

**Comment by jlblancoc** [comment]
Hi @rachit173 !

> Also I tried to install ROS(kinetic-kame) on 18.04 but there seems to be no version available, is there a workaround?

ROS melodic will be available within a few weeks (during May 2018).

I agree with @jolting 's view of making the web interface independent of ROS (although it could be later integrated as a ROS node) for the sake of portability; also, using all format options should be considered for the transport layer: binary blobs, JSON,... The former will probably be much more efficient, so in case of using JSON, there should be good reasons to pick that choice.
Cheers.

**Comment by rachittibrewal** [comment]
Hi
My exams are almost over, I have one easy paper left for May 9. So I can start with work. Should I start with the first phase work?

**Comment by jolting** [comment]
Sure. That's the occupancy grid task right?

**Comment by rachittibrewal** [comment]
It is the publisher/subscriber mechanism for MRPT.

**Comment by jolting** [comment]
Cool. You may begin coding any time you wish.

**Comment by jolting** [comment]
@rachit173
Do you have some time in the next few days to have a quick video chat to discuss your project?

I'm in California. I presume you're in India, so your morning will probably works best for me.

**Comment by rachittibrewal** [comment]
I am free tomorrow morning any time from 8:00 Indian Standard Time(GMT+5:30). Any particular hour?

On May 10, 2018 10:03, Hunter Laux <notifications@github.com> wrote:

@rachit173<https://github.com/rachit173>
Do you have some time in the next few days to have a quick video chat to discuss your project?

I'm in California. I presume you're in India, so your morning will probably works best for me.

**Comment by rachittibrewal** [comment]
@jolting @jlblancoc 
For the publisher subscriber pattern , I am confused between brokered systems(like ROS with the roscore or master) or brokerless system. In ROS the broker accepts all the publishers messages and distributes the messages to subscribers requesting the message. The master can also be used for RPC. So should i continue with a  brokered system like ROS or are there some other design patterns which would be more advantageous for our purpose of web apps?

**Comment by jolting** [comment]
ROS core doesn't distribute the messages. It acts like a name service allowing nodes to connect with each other, but the nodes do talk directly with each other for performance reasons. Messages aren't relayed.

ZeroMQ has a good description of broker vs brokerless.
http://zeromq.org/whitepapers:brokerless

This project isn't intended to reinvent ROS. There are two actors in this system. A web browser and an MRPT server.

A single point to point data channel such as a websocket doesn't need a broker. You can still implement pub/sub at the protocol layer.

See ethereum:
https://github.com/ethereum/go-ethereum/wiki/RPC-PUB-SUB

**Comment by rachittibrewal** [comment]
cppethereum is great but it does not have websocket support unlike goeth, but basic code of cppethereum rpc with beast library could solve this.
Also I have created mrpt-server personal repo, I will be pushing code to it and have started with basic configuration file till I get a clarity of the RPC. Should I continue with my repo or can you create a new repo for mrpt-server.

**Comment by jolting** [comment]
Ethereum is just an example of RPC. I would not suggest adding that as a dependency.

I prepared this template a while back:
https://github.com/MRPT/mrpt-web

You can fork that or start your own.

**Comment by rachittibrewal** [comment]
Update:
forked mrpt-web.
created a sync web server for ws using beast 
The ws connection works in chrome but firefox does not.
For firefox, the site says that SSL is required thus, I will create a separate app for SSL.
Appart from synchronois, asynchronous and coroutine ws server, much of the code is present in beast documentation.
I am thinking of placing the different servers in CWebSocketServer.h file

**Comment by jolting** [comment]
Ok. You can send a pull request to the mrpt-web repo when you want to review your changes.

**Comment by rachittibrewal** [comment]
@jolting @jlblancoc  I have implemented jsonrpccpp using websockets and pushed it to my fork. The code is based mostly on cppethereum and works with jsonrpc-2.0 js client (I tried with plain websockets). However, it currently does not support SSL and thus does not work with firefox sometimes. I am currently working on it. Currently the file implementing websocket server is CWebSocketJsonRpcServer.h but I suspect that it is not doing a clean exit and thus if I immediately run the server on the same port it gives errors but after sometime it works fine. Any pointers regarding that? 

Regarding dependencies, beast has a header files in beast/experimental which were giving error when I removed the included files.

Also regarding licenses and credits, I have copied lot of code from cppethereum, should i just include their license or put it at start of every code file used?

**Comment by jolting** [comment]
>I have implemented jsonrpccpp using websockets and pushed it to my fork. The code is based mostly on cppethereum and works with jsonrpc-2.0 js client (I tried with plain websockets). 

To avoid licensing issue you should rewrite the sections you based on cppethereum. Code derived from GPL code cannot be BSD code.

>Regarding dependencies, beast has a header files in beast/experimental which were giving error when I removed the included files.

You should use beast as it is shipped with boost 1.66 or later.

**Comment by rachittibrewal** [comment]
Update : Pushed websocket with SSL.
Reading up on rosbridge protocol
It has the following operations:
- advertise
- unadvertise
- subscribe
- unsubscribe
- call_service
- advertise_service
- unadvertise_service
- service_request
- service_response
I was also thinking about implementing a similar protocol. Are there changes required to protocol which will
make it more efficient ? Can you give me a brief idea about what kind of design would be good design

**Comment by jolting** [comment]
Minimalism is key here.

Think of a way to version the protocol. There a reason why MRPT can still read old data files despite all the changes over the years. Backward compatibility requires some thought. I would start with some way of versioning the protocol.

Services are actually ROS's way of doing RPC Request/Reply. I would look at that next.

The following are related to ROS's pub/sub
```
advertise
unadvertise
subscribe
unsubscribe
```
I certainly wouldn't start with Pub/Sub.

**Comment by rachittibrewal** [comment]
check this link for mrpt-web protocol
https://github.com/rachit173/mrpt-web/wiki
I have added the protocol associated with mrpt-web, method call protocol has already been implemented. We can decide the protocol in the wiki and it will also serve as basic documentation. Please suggest changes in protocol there. 
Also I am confused about converting different message or parameters like costmap, point, laserscan data type to json fields for the transport. Should i use ROS's message type as they already have this available or some other package.
Once this conversion package is done, I will create a basic app to demonstrate the method call protocol.

**Comment by jolting** [comment]
>I have added the protocol associated with mrpt-web, method call protocol has already been implemented. We can decide the protocol in the wiki and it will also serve as basic documentation. Please suggest changes in protocol there.

Awesome. Documentation is key.

> Also I am confused about converting different message or parameters like costmap, point, laserscan data type to json fields for the transport. Should i use ROS's message type as they already have this available or some other package.

This should be done for MRPT data structures.

I think it should be possible to create something like CArchive for JSON.
https://github.com/MRPT/mrpt/blob/master/libs/serialization/include/mrpt/serialization/CArchive.h

**Comment by jlblancoc** [comment]
👍 for starting documenting things since the beginning in a wiki. Keep that!

> I think it should be possible to create something like CArchive for JSON.
> https://github.com/MRPT/mrpt/blob/master/libs/serialization/include/mrpt/serialization/CArchive.h

Hmmm... After this GSOC project, there will be THREE different serialization protocols:
* Via CArchive `<<` & `>>` operators. See [docs](http://mrpt.ual.es/reference/master/group__mrpt__serialization__grp.html).
* MEXPLUS + `writeToMatlab()` to (de)serialize to MATLAB. See [example](https://github.com/MRPT/mrpt/blob/80c44e1114fb113b31b56577eec7b80fb48402e6/libs/obs/src/CObservationImage.cpp#L101-L124). 
* Whatever is done to handle JSON.

It would be fantastic if we could somehow centralize all serialization methods in one single place, as far as it may be possible without breaking what currently works.

A first simple idea (that needs more thinking, I don't mean it's optimal) would be adding new methods to CArchive to `writeXXX()` and `readXXX()`  for `XXX=mex|json`, for example. The methods should accept variables and their "name". 
Another place to look for unification, is [the triplet of virtual methods](https://github.com/MRPT/mrpt/blob/master/libs/serialization/include/mrpt/serialization/CSerializable.h#L39-L59) of CSerializable. I'm not sure if it would be feasible (it depends on whether JSON serializations require the variable names in addition to their values), but it would be fantastic to allow existing code to automatically work if the `CArchive` that is passed had some "flag" to reflect which serialization protocol is desired: binary or JSON (MEX is so different that I **don't think** we have chances of unifying it here...). 

Notice that CArchive is the place where "serialization" happens. That class actually is in charge of **interpreting** data to/from a text or binary stream, which may be an underlying mrpt::io::CStream or a std::stream.

**Comment by jolting** [comment]
> it depends on whether JSON serializations require the variable names in addition to their values

Yeah JSON needs a schema.

A solution could look like this:
```
template <typename SCHEMA_CAPABLE>
SCHEMA_CAPABLE CPose2D::serializeTo() const
{
  SCHEMA_CAPABLE out; 
  out["x"]    = m_coords[0];
  out["y"]    = m_coords[1];;
  out["phi"] = m_phi;
  return out;
}
```

Then you can do.
```
  CPoint2D point;
  auto json_point = point.serializeTo<Json::Value>();
```

A little bit more complicated version would be

```
template <typename SCHEMA_CAPABLE>
SCHEMA_CAPABLE SomeObject::serializeTo() const
{
  SCHEMA_CAPABLE out;
  out["nested"]    = m_somenested.serializeTo();
  return out;
}
```

Clearly, this could get a little bit more declarative, but might as well start with something simple.

**Comment by rachittibrewal** [comment]
I am facing the following error
```
CMakeFiles/mrpt-ws-rpc.dir/main.cpp.o: In function `mrpt::serialization::CSerializable::~CSerializable()':
main.cpp:(.text._ZN4mrpt13serialization13CSerializableD2Ev[_ZN4mrpt13serialization13CSerializableD5Ev]+0xf): undefined reference to `vtable for mrpt::serialization::CSerializable'
CMakeFiles/mrpt-ws-rpc.dir/main.cpp.o: In function `mrpt::poses::CPoint2D::operator delete(void*)':
main.cpp:(.text._ZN4mrpt5poses8CPoint2DdlEPv[_ZN4mrpt5poses8CPoint2DdlEPv]+0x14): undefined reference to `mrpt::aligned_free(void*)'
CMakeFiles/mrpt-ws-rpc.dir/main.cpp.o: In function `mrpt::serialization::CSerializable::CSerializable()':
main.cpp:(.text._ZN4mrpt13serialization13CSerializableC2Ev[_ZN4mrpt13serialization13CSerializableC5Ev]+0x1b): undefined reference to `vtable for mrpt::serialization::CSerializable'
CMakeFiles/mrpt-ws-rpc.dir/main.cpp.o: In function `mrpt::poses::CPoint2D::CPoint2D(double, double)':
main.cpp:(.text._ZN4mrpt5poses8CPoint2DC2Edd[_ZN4mrpt5poses8CPoint2DC5Edd]+0x36): undefined reference to `vtable for mrpt::poses::CPoint2D'
main.cpp:(.text._ZN4mrpt5poses8CPoint2DC2Edd[_ZN4mrpt5poses8CPoint2DC5Edd]+0x44): undefined reference to `vtable for mrpt::poses::CPoint2D'
CMakeFiles/mrpt-ws-rpc.dir/main.cpp.o: In function `CRPCRawLog::Playlist_GetItems()':
main.cpp:(.text._ZN10CRPCRawLog17Playlist_GetItemsEv[_ZN10CRPCRawLog17Playlist_GetItemsEv]+0x43): undefined reference to `Json::Value mrpt::poses::CPoint2D::serializeTo<Json::Value>() const'
CMakeFiles/mrpt-ws-rpc.dir/main.cpp.o: In function `mrpt::rtti::TRuntimeClassId const* mrpt::rtti::CLASS_ID_impl<mrpt::rtti::CObject>()':
main.cpp:(.text._ZN4mrpt4rtti13CLASS_ID_implINS0_7CObjectEEEPKNS0_15TRuntimeClassIdEv[_ZN4mrpt4rtti13CLASS_ID_implINS0_7CObjectEEEPKNS0_15TRuntimeClassIdEv]+0x5): undefined reference to `mrpt::rtti::CObject::GetRuntimeClassIdStatic()'
CMakeFiles/mrpt-ws-rpc.dir/main.cpp.o: In function `mrpt::poses::CPoint2D::~CPoint2D()':
main.cpp:(.text._ZN4mrpt5poses8CPoint2DD2Ev[_ZN4mrpt5poses8CPoint2DD5Ev]+0xf): undefined reference to `vtable for mrpt::poses::CPoint2D'
main.cpp:(.text._ZN4mrpt5poses8CPoint2DD2Ev[_ZN4mrpt5poses8CPoint2DD5Ev]+0x1d): undefined reference to `vtable for mrpt::poses::CPoint2D'
collect2: error: ld returned 1 exit status
apps/mrpt-ws-rpc/CMakeFiles/mrpt-ws-rpc.dir/build.make:105: recipe for target 'apps/mrpt-ws-rpc/mrpt-ws-rpc' failed
make[2]: *** [apps/mrpt-ws-rpc/mrpt-ws-rpc] Error 1
CMakeFiles/Makefile2:140: recipe for target 'apps/mrpt-ws-rpc/CMakeFiles/mrpt-ws-rpc.dir/all' failed
make[1]: *** [apps/mrpt-ws-rpc/CMakeFiles/mrpt-ws-rpc.dir/all] Error 2
Makefile:129: recipe for target 'all' failed
make: *** [all] Error 2
```
<b>main.cpp</b>
```cpp
#include <jsonrpccpp/server/abstractserverconnector.h>

#include <mrpt/poses/CPoint2D.h>
#include <mrpt/serialization/CSerializable.h>
#include <mrpt/rtti/CObject.h>
#include <mrpt/web/CWebSocketUtility.hpp>
#include <mrpt/web/CWebSocketAdvanced.h>
#include <mrpt/web/CModularServer.h>

#include <CRPCRawLogFace.h>

#include <cstdlib>
#include <functional>
#include <thread>
#include <string>
#include <mutex>
#include <memory>

using namespace mrpt::poses;
class CRPCRawLog : public CRPCRawLogAbstract
{
public:
  virtual RPCModules implementedModules() const override
  {
    return RPCModules{RPCModule{"CRPCRawLog","1.0"}};
  }
  Json::Value Playlist_GetPlaylists() override
  {
    Json::Value ch;
    ch[0]["name"] = "Jabberwock";
    ch[0]["chapter"] = 1;
    ch[1]["name"] = "Cheshire Cat";
    ch[1]["chapter"] = 6;
    ch[2]["name"] = "Mad Hatter";
    ch[2]["chapter"] = 7;

    // create the main object
    Json::Value val;
    val["book"] = "Alice in Wonderland";
    val["year"] = 1865;
    val["characters"] = ch;
  return ch;
  }
  Json::Value Playlist_GetItems() override
  {
   //Added this code
    CPoint2D point;
    auto jsonPoint = point.serializeTo<Json::Value>();
    return jsonPoint;
  }
};
```
CPoint2D.cpp
```cpp
template <typename SCHEMA_CAPABLE>
SCHEMA_CAPABLE CPoint2D::serializeTo() const
{
	SCHEMA_CAPABLE out;
	out["x"] = m_coords[0];
	out["y"] = m_coords[1];
	return out;
}
```
Also since templatized virtual functions are not allowed, how to add it to CSerializable.h

**Comment by rachittibrewal** [comment]
The above error was due to not linking the libraries properly, that is solved.
I am thinking of adding this serialization method to MRPT data structures as you have suggested, in the form of templatized serializeTo function. 
Adding the message type could be useful for js library later like,
```cpp
out["datatype"] = "CPose2D";
```

**Comment by jolting** [comment]
That seems reasonable. Also, this is important for serializeFrom.
You might want to add a version as well.

**Comment by rachittibrewal** [comment]
I will adding the JSON schemas in this wiki and have added some Schema for poses.
https://github.com/rachit173/mrpt-web/wiki/JSON-serialization-of-MRPT-Data-Structures
Also,
```cpp
/** Templatized serializeTo function */
template <typename SCHEMA_CAPABLE>
SCHEMA_CAPABLE serializeTo() const
{
	SCHEMA_CAPABLE out;
	out["datatype"] = this->GetRuntimeClass()->className;
	out["version"] = 1;
	out["x"] = m_coords[0];
	out["y"] = m_coords[1];
	return out;	
}

/** Templatized serializeFrom function 
 * Serializes only if the datatype matched to className 
*/
template <typename SCHEMA_CAPABLE>
void serializeFrom(SCHEMA_CAPABLE& in)
{
	uint8_t version = in.get("version",0);
	if(in["datatype"] == this->GetRuntimeClass()->className)
	{
		switch(version)
		{
			case 1:
			{
				m_coords[0] = in["x"];
				m_coords[1] = in["y"];
			}
			break;
			default:
				MRPT_THROW_UNKNOWN_SERIALIZATION_VERSION(version)
		}
	}
}
```
Are these serialization functions fine?

**Comment by rachittibrewal** [comment]
Update:
1. Added cmake file for making changes to stub generated by jsonrpc for use by our project. One can generate stub (slight modification of jsonrpc stub) by specifying the json specification and using <b>jsonrpcstub_create</b> macro in cmake.

2. The remote procedure call can be generated as and when required by adding the specification json, generating stub and adding its object to CModularServer.

3. I need to use RPC to create Publish / Subscribe advertisements as well as calls. For this, the protocol has been described in this [wiki](https://github.com/rachit173/mrpt-web/wiki).

4. Since we have not developed any Pub/Sub mechanism, for the purpose of creating the web apps later, should I use ros_master api and build a wrapper around it for Publishing and Subscribing to topics. This wrapper will be based on the stub generated in this [code](https://github.com/rachit173/mrpt-web/blob/master/apps/mrpt-ws-rpc/main.cpp). Then I can use mrpt libraries along with ROS Pub/Sub for server side code of the apps.

I am currently working on serialization and will start work on the RPC Pub/Sub wrapper once decided.

**Comment by rachittibrewal** [comment]
ROS's pub/sub
```
advertise
unadvertise
subscribe
unsubscribe
```
I had started working on the above, the main issue here is keeping track of clients (for sending messages from server to client without a signal).  Currently I have created a store class with a map(map<string,shared_ptr<connection>>) and plan to use the string as the unique identifier for the client.
So the subscribe message would look llike,
```
{
"method":Subscriber.subscribe
"params":
{
 "clientID": key,
 "topic" : topic,
 "throttle_rate"......
}
}
```

Now the subscribe method in the derived class of the stub will have access to the store and can use this to add the connection object to corresponding topic in a SubscribeManager class instance.
Similarly the connection object will contain the new subscribed topic.

A good decision of the key may be needed to address security issues. I was thinking of sending a randomly generated unique identifier to the client on connection and using that identifier as the key.

This is a vague, I am currently working on the code. But wanted your suggestions. The main issue facing here is keeping track of client connection for sending messages from subscribed topics to corresponding client.

**Comment by rachittibrewal** [comment]
I am developing the stubs related to the above in [dev-pubsub](https://github.com/rachit173/mrpt-web/tree/dev-pubsub). Also I have tried to clean up the large number of commits into much smaller number of commits depicting the logical sequence of development.

**Comment by rachittibrewal** [comment]
@jolting 
I have removed dependancy from jsonrpccpp by adding self written RPC handler. Currently, I am developing it and the RPC PubSub in [dev-rpc-pubsub](https://github.com/rachit173/mrpt-web/tree/dev-rpc-pubsub).
The changed protocol has been added to the [wiki](https://github.com/rachit173/mrpt-web/wiki).

**Comment by jolting** [comment]
Avoid mixing common.js style imports and es6 style imports. ES6 imports are preferred.

I'm wondering if RxJS could be used in lieu of EventEmitter2. What do you think?

For unit testing do you have a preference for mocha?
Jest seems a little more popular and I have more experience with it.

Good work so far on the JS stuff, so far. I'm excited to see something fully working and demonstrable.

**Comment by rachittibrewal** [comment]
I am looking into RxJS, maybe at a more mature stage of library,using it would be good. I don't have a preference for mocha(it seems easy to use), I will learn jest and change the unit test to it.
I am facing an issue currently, the bundled js generated by webpack seems to have errors, is it common, I am trying to debug it but any leads would be good.

**Comment by jolting** [comment]
Looks like you're trying to use the `ws` module in client-side code.

Here's an example how one library solved it:
https://github.com/maxogden/websocket-stream/blob/3f39dcbc098f661ee12d7deba297b1420a0a0e07/package.json#L43

https://github.com/maxogden/websocket-stream/blob/3f39dcbc098f661ee12d7deba297b1420a0a0e07/ws-fallback.js

**Comment by rachittibrewal** [comment]
That works :).

**Comment by rachittibrewal** [comment]
Update:
Developed modules for:
* Setup 3D scene, add meshes for arrow, markers and Pointcloud.
* PointCloud is based on serialization of CPointCloud, should the colorFromDepth parameter also be used for the pointcloud creation?
* I am currently planning to make some more meshes of basic shapes, OccupancyGrid, laserscan.
![screenshot from 2018-06-26 00-03-24](https://user-images.githubusercontent.com/32801794/41868773-b0299964-78d4-11e8-8721-70b2a03c5d74.png)
![screenshot from 2018-06-26 00-14-19](https://user-images.githubusercontent.com/32801794/41869245-063746de-78d6-11e8-9626-226ce5149a77.png)

**Comment by jolting** [comment]
Good work.

It would be nice to make coloring the clouds a configurable parameter in javascript. Common coloring parametizations include axis, depth and flat color. Of course if the point cloud has color then it should be an option. I find it useful to switch between these when trying to understand data.

I don't think colorFromDepth is a necessary parameter.

@jlbancoc care to share any insight?

**Comment by jlblancoc** [comment]
Hi guys,

Ok, I'll take a look at this tomorrow (CEST).

As a newbie regarding serious js development, did you already included some
detailed instructions in your fork on how to launch some demo(s)? If not,
please add it to the readme/wiki/wherever and I'll beta test it ;-)

Cheers

**Comment by rachittibrewal** [comment]
[wiki](https://github.com/rachit173/mrpt-web-js/wiki) added.

**Comment by jlblancoc** [comment]
Thanks!

Just a minor edit to the wiki... there's a missing "cd":

> git clone https://github.com/rachit173/mrpt-web-js.git
> cd mrpt-web-js
> npm install


However, I found errors during "npm install". Apparently, a missing dependency (?). 

```
 ├─┬ schema-utils@0.4.5 
│ │ └─┬ UNMET PEER DEPENDENCY ajv@6.5.1

...

npm WARN optional Skipping failed optional dependency /chokidar/fsevents:
npm WARN notsup Not compatible with your operating system or architecture: fsevents@1.2.4
npm WARN optional Skipping failed optional dependency /sane/fsevents:
npm WARN notsup Not compatible with your operating system or architecture: fsevents@1.2.4
npm WARN optional Skipping failed optional dependency /watchpack/chokidar/fsevents:
npm WARN notsup Not compatible with your operating system or architecture: fsevents@1.2.4
npm WARN ajv-keywords@3.2.0 requires a peer of ajv@^6.0.0 but none was installed.
npm WARN mrpt-web-js@1.0.0 No repository field.
npm ERR! Linux 4.4.0-128-generic
npm ERR! argv "/usr/bin/nodejs" "/usr/bin/npm" "install"
npm ERR! node v4.2.6
npm ERR! npm  v3.5.2
npm ERR! file sh
npm ERR! code ELIFECYCLE
npm ERR! errno ENOENT
npm ERR! syscall spawn

npm ERR! babel-env@2.4.1 postinstall: `node ./bin/babel-env`
npm ERR! spawn ENOENT
npm ERR! 
npm ERR! Failed at the babel-env@2.4.1 postinstall script 'node ./bin/babel-env'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the babel-env package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node ./bin/babel-env
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs babel-env
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls babel-env
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /home/jlblanco/code/mrpt-web-js/npm-debug.log
```

**Comment by jolting** [comment]
>npm ERR! node v4.2.6
>npm ERR! npm  v3.5.2

The 4 branch of node is EOL.

Typically you'll want to install a new version of node:
https://nodejs.org/en/download/package-manager/

It's also typical to upgrade npm. You can use npm.
`npm i -g npm@latest`

You'll need root priviledges unless you set `.npmrc` to use a different node prefix.

That might not be the problem, but it's a good place to start.

**Comment by jlblancoc** [comment]
Indeed, that was the problem! :-)
Tests now run smoothly. 

On [the wiki](https://github.com/rachit173/mrpt-web-js/wiki/Scene-and-Meshes): "will created in the dist folder after running build". Please, provide the direct command for building (as a beginner with nodejs, all these steps were not trivial to me ;-): something like: `npm run-script build`

I was able to build the js lib and run the basic 3D html demo, good!

Now, on the original question: 

> It would be nice to make coloring the clouds a configurable parameter in javascript. Common coloring parametizations include axis, depth and flat color. Of course if the point cloud has color then it should be an option. I find it useful to switch between these when trying to understand data.

I completely agree with Hunter here. Size of points should be configurable as well. 
Color from height is not a bad idea of something to have, since it's useful when handling large pointclouds, but it's a low priority feature. 

I would say it's far more important to add: 
* occupancy grid maps
* XYZ corner markers, like MRPT's [mrpt::opengl::stock_objects::CornerXYZSimple](http://mrpt.ual.es/reference/master/namespacemrpt_1_1opengl_1_1stock__objects.html#abf6df2ba6333fb2ec8f33b9f9a47bb7a), including the two parameters for line width and line length. 

PS @jolting : what do you think about adding some example HTML pages to the repo with examples? Would it be "good practices"? (It would require building the package first...)

PS2: What do you think about adding a "bigger picture" graph / schematic (could be done even in PowerPoints / LibreOffice Impress) with the role of each repo in this project?: MRPT, mrpt-web, mrpt-web-js,... (user apps?)

**Comment by rachittibrewal** [comment]
Surely, a bigger picture is good for the project understanding. I will plan and make it.
I was planning to simultaneously start with web app version of RawLogViewer. So, in which repository should this app belong?
For the C++ part of the app, I am thinking of adding that to mrpt-web repo.
Also, which are the core features of RawLogViewer which should be started with?

About RawLogViewer App: Through this app I will basically demonstrate how instead Qt or wx gui, we can use Web UI with RPC by creating stubs for required functions. Is the basic theme correct?

**Comment by jolting** [comment]
Examples are good.

It's common after libraries have been made unopinionated to create VueJS or React component libraries around them. Many times both a ReactJS and a VueJS version are created because the war is very much on.

https://hasvuepassedreactyet.surge.sh/
Star VueJS please. 👍 

Once they've been wrapped in components it tends to make reuse quite a bit easier and examples based on components might be more useful in the long run.

**Comment by rachittibrewal** [comment]
Update:
1. Pushing code related to rawlog-web-ui backend (C++ server) in [mrpt-web](https://github.com/rachit173/mrpt-web/tree/master).
2. Pushing code related to frontend of rawlog-web-ui to [rawlog-web-ui](https://github.com/rachit173/rawlog-web-ui).
The frontend node package depends on mrpt-web-js package and thus one needs to link those using npm link. Instructions can be found in the repo's README.
I think this will be required until the mrpt-web-js is published to npm.
@jolting is there a simpler method for this.
Should I continue with creating separate repos for the apps?
Also, I was not clear on what type of components should be created? Could you give some examples. Currently, I am creating components in the app for use there. I was thinking of creating this app before starting component development, so that I can get a better idea about what kind of components are useful in apps.

P.S. I have done some major restructuring to mrpt-web dev-rpc-pubsub is now master.

**Comment by jolting** [comment]
npm link the right thing to use for development. Eventually we'll publish something under `@mrpt/<some lib name>`. Feel free to come up with a better name.

**Comment by rachittibrewal** [comment]
```cpp
    if (classID == CLASS_ID(CActionRobotMovement2D))
    {
      // ----------------------------------------------------------------------
      //              CActionRobotMovement2D
      // ----------------------------------------------------------------------
      cs.changeSelection(1);

      CActionRobotMovement2D::Ptr act =
        std::dynamic_pointer_cast<CActionRobotMovement2D>(sel_obj);

      CPose2D Ap;
      CMatrixDouble33 mat;
      act->poseChange->getCovarianceAndMean(mat, Ap);

      ss << "Robot Movement (as a gaussian pose change):\n";
      ss << " Mean = " << Ap << endl;

      ss << format(" Covariance:     DET=%e\n", mat.det());

      ss << format(
        "      %e %e %e\n", mat(0, 0), mat(0, 1), mat(0, 2));
      ss << format(
        "      %e %e %e\n", mat(1, 0), mat(1, 1), mat(1, 2));
      ss << format(
        "      %e %e %e\n", mat(2, 0), mat(2, 1), mat(2, 2));

      ss << endl;

      ss << " Actual reading from the odometry increment = "
          << act->rawOdometryIncrementReading << endl;

      ss << format(
        "Actual PDF class is: '%s'\n",
        act->poseChange->GetRuntimeClass()->className);

      if (act->poseChange->GetRuntimeClass() ==
        CLASS_ID(CPosePDFParticles))
      {
        CPosePDFParticles::Ptr aux =
          std::dynamic_pointer_cast<CPosePDFParticles>(
            act->poseChange.get_ptr());
        ss << format(
          " (Particle count = %u)\n",
          (unsigned)aux->m_particles.size());
      }
      ss << endl;

      ss << "Estimation method: ";
      switch (act->estimationMethod)
      {
        case CActionRobotMovement2D::emOdometry:
          ss << "emOdometry\n";
          break;
        case CActionRobotMovement2D::emScan2DMatching:
          ss << "emScan2DMatching\n";
          break;
        default:
          ss << "(Unknown ID!)\n";
          break;
      };

      // Additional data:
      if (act->hasEncodersInfo)
      {
        ss << format(
          " Encoder info: deltaL=%i deltaR=%i\n",
          act->encoderLeftTicks, act->encoderRightTicks);
      }
      else
        ss << "Encoder info: Not available!\n";

      if (act->hasVelocities)
      {
        ss << format(
          " Velocity info: v=%s\n",
          act->velocityLocal.asString().c_str());
      }
      else
        ss << "Velocity info: Not available!\n";

      // Plot the 2D pose samples:
      unsigned int N = 1000;
      vector<CVectorDouble> samples;
      vector<float> xs(N), ys(N), ps(N), dumm(N, 0.1f);

      // // Draw a set of random (x,y,phi) samples:
      // act->poseChange->drawManySamples(N, samples);

      // // Pass to vectors and draw them:
      // for (unsigned int i = 0; i < N; i++)
      // {
      //   xs[i] = samples[i][0];
      //   ys[i] = samples[i][1];
      //   ps[i] = RAD2DEG(samples[i][2]);
      // }

      // lyAction2D_XY->SetData(xs, ys);
      // lyAction2D_PHI->SetData(ps, dumm);

      // plotAct2D_XY->Fit();
      // plotAct2D_PHI->Fit();
      /** Serialize xs, ys, ps, dumm data
       *
       * Code Here
       */
    }
```
this line of code is causing an error 
```cpp
     act->poseChange->drawManySamples(N, samples);
```
```
double free or corruption (out)
```
I checked the type of poseChange and it is 	mrpt::containers::deepcopy_poly_ptr<mrpt::poses::CPosePDF::Ptr>

@jolting @jlblancoc 
Is this error due to the deepcopy_poly_ptr type?
I had copied the code from RawlogViewer app, and made certain changes to it.

**Comment by rachittibrewal** [comment]
I have added project boards, for more efficient completion of app and library, please tell in case any improvements to project.
[board1](https://github.com/rachit173/mrpt-web-js/projects/1)
[board2](https://github.com/rachit173/rawlog-web-ui/projects/1)

**Comment by jlblancoc** [comment]
> Is this error due to the deepcopy_poly_ptr type?

I just tested the equivalent code inside RawLogViewer and it works without crashing. 
Please, could you put together a small .cpp file + its CMakeLists.txt that reproduces the crash? If yes, open an issue on the main MRPT/mrpt repo with that test code.

**Comment by jolting** [comment]
Minor critique of the code so far. I find mapMutation and mapAction cleans up the code quite a bit.

**Comment by rachittibrewal** [comment]
I was not aware of the better style, I will make the change straight away.

**Comment by rachittibrewal** [comment]
I have added the basic ideas about the library in the google slides presentation, it will be updated as the work progresses further. The slide aims to explain the role of each of the libraries.
[slides](https://docs.google.com/presentation/d/1UvH86SG-KHGqXmseFJNyZ1oUNbfNUdS14QyvK6kmkYA/edit?usp=sharing)

**Comment by jlblancoc** [comment]
Hunter will probably take a deeper look to your recent changes, but here comes a quick overview to the slides.
Minor typos (I think) and change requests:
*  "It comprises of:" -> "It comprises:"  
* "Mrpt..." -> "mrpt" or "MRPT" looks better, doesn't it?
* Class names (like "CWebSocket", etc.): it would be great to have those names linked (href) to some documentation for the classes, or at least, to the .h of the corresponding class. Ideally, the former... I don't know how easy is to setup a Doxygen job in a readthedocs site or alike? Also, one can always just use github pages. See [this example](https://github.com/jlblancoc/nanoflann/blob/master/.travis.sh#L24-L62), published [here](http://jlblancoc.github.io/nanoflann/).
* Adding a schematic graph with boxes, arrows, etc. would be helpful. At the highest level (mrpt-web, mrpt-web-js) and perhaps also with the relationship between some of the C++ classes. Just thinking aloud... but I'm quite a visual guy: an image is normally worth 1000 words ;-)
* It would be "educative" to show an example of how some (small!) piece of data is represented as JSON, showing the native C++ class name, etc.

The RawLogViewer app looks impressive! Again... it would be great to provide detailed instructions on how to launch it (perhaps a link to a wiki page?).

**Comment by rachittibrewal** [comment]
Thanks @jlblancoc
The changes requested seem necessary, I will work  on them.

@jolting how should the documentation of mrpt-web-js proceed, should I do in GitHub pages, wiki or what is the popular practice?

**Comment by jlblancoc** [comment]
> The changes requested seem necessary, I will work on them.

Good, thanks!

> For the C++ documentation, I would like to use doxygen but am unaware of it, can you provide some pointers about it( or some links). 

You can start installing doxygen (apt-get), and using its [-g](https://www.stack.nl/~dimitri/doxygen/manual/doxygen_usage.html) parameter to create a brand new config file. Browse it and the Doxygen docs to see the meaning of the most important parameters. 
Mostly, the `INPUT` key which instruct doxygen where to search for source code. 

The `Doxyfile` file could be created just on the root directory. 

Then, just invoking `doxygen` will generate an HTML report of all C++ classes. You can disable all other non-HTML generators (LaTeX, RTF, etc.). As you browse the generated HTML you'll discover plenty of ways of improving it ;-)

Take a look at MRPT's [Doxyfile](https://github.com/MRPT/mrpt/blob/master/doc/doxygen_project.txt.in) template file (although it's overcomplicated for your project needs)

> However, I prefer to push that a little towards the end.

OK, no rush for this second evaluation.

**Comment by jolting** [comment]
I forked your repo
https://github.com/MRPT/mrpt-web-js
I added a circleci config.

Looks like the unit tests hangs there after it finishes.

https://circleci.com/gh/MRPT/mrpt-web-js/2

I also changed the package name so that it begin with `@mrpt`. I previously registered that for the MRPT org.

**Comment by jolting** [comment]
>@jolting how should the documentation of mrpt-web-js proceed, should I do in GitHub pages, wiki or what is the popular practice?

For documenting javascript libraries, the conventional way is a fancy webpage that shows off your web dev skills. A wiki is a good place to start though. I'd rather see finished well documented code rather than a fancy webpage. You can always add that later.

If you do plan on throwing up a webpage quickly for either documentation or demo give surge.sh a shot.
https://surge.sh/

**Comment by rachittibrewal** [comment]
Since you have created MRPT/mrpt-web-js,rawlog-web-ui, should I now fork them or can I continue using my existing repos and create pull requests. I had setup project boards there so, it would be useful, if I could continue to develop on those repos :-)

**Comment by jolting** [comment]
Keep working on your board. I was just trying to get a build pipeline for npmjs.com

**Comment by rachittibrewal** [comment]
Thanks for the evaluation.
I had added plotly.js, since two.js did not provide direct graphing and plotting. I was planning on completing the rawlog-web-ui and improving the mrpt-web-js library by adding modules, tests and documentation. This would be the plan for next week. After that what should be my focus, another app, or examples? I plan to keep the last week for documentation and unit tests, so, I will have 10 days for either an app or examples.

**Comment by jolting** [comment]
If you can produce a youtube video + documentation on how to set everything up from scratch it would be great. The target audience is C++ programmers who don't know how to setup node or javascript.

After that more apps or keep adding capabilities to the raw log viewer.

**Comment by jlblancoc** [comment]
>   The target audience is C++ programmers who don't know how to setup node
or javascript.

I'll be an archetypal example of those guys, so will provide you valuable
feedback on your tutorial ;-)

**Comment by jolting** [comment]
Here's an example of what I consider a good video tutorial format:
https://www.vuemastery.com/courses/intro-to-vue-js

**Comment by jolting** [comment]
If you get done with the rawlog viewer, the documentation and video tutorials. Then you can work on the Reactive Navigation demo.

**Comment by rachittibrewal** [comment]
I will start documentation, build configuration of everything.

**Comment by rachittibrewal** [comment]
Started with something basic:
ESDoc for automated build of source documentation and published it on http://mrpt-web.surge.sh/
thinking of adding tutorials and further examples here itself
added CI for docs on branch [dev-docs](https://github.com/rachit173/mrpt-web-js/tree/dev-docs)
Improved docs, added examples (code snippets) to many of the classes.

**Comment by jlblancoc** [comment]
> ESDoc for automated build of source documentation and published it on http://mrpt-web.surge.sh/

👍 looks great!

Shouldn't that also include: 
* Short instructions/(link to)examples on how to use those classes in real apps?
* Instructions to launch the RawlogViewer demo app & a link to the most relevant parts of the code?

**Comment by rachittibrewal** [comment]
Thanks, I am going to add tutorials to the manual section

**Comment by jolting** [comment]
Looks like the mrpt-web-js unit tests are still broken.

**Comment by jolting** [comment]
What do you think about specifying a directory on the server in which rawlogs will be served up from?

It is a bit awkward to have to know the name of the file on the file system.

**Comment by jolting** [comment]
I think there needs to be a method for connecting to a server other than localhost:5000.

**Comment by jolting** [comment]
For the 3d animation is there any way to move the pivot point?

**Comment by jolting** [comment]
One thing to note about the 3d scan visualization is that it is capable of rendering multiple sensors. Each sensor will have its own sensor label and have a different sensor height.

**Comment by jolting** [comment]
Let me know if the 2D plots work. I'm excited to try them out.

**Comment by rachittibrewal** [comment]
I have added address bar for specifying the websocket address and port, 2d plots works in map and path generation component and will soon work for individual rawlog (debugging it).

For the rawlog address, how do you feel about sending file from client to server using websocket (we can then use file dialog box).  If we wish to go with directory, then should I create a specific directory from which the server will give the list of files available for visualisation from the directory?

3d animation currently uses trackballcontrols, for changing the pivot the press the right click and move the mouse. There are other controls out there, https://threejs.org/examples/?q=controls#misc_controls_map, but have their own pros and cons.

For 3d scan visualisation, are you aware of some rawlog with multiple sensor data. Then I would check how it functions with the RawLogViewer app and modify the app for multiple sources.

**Comment by jolting** [comment]
> For the rawlog address, how do you feel about sending file from client to server using websocket (we can then use file dialog box). If we wish to go with directory, then should I create a specific directory from which the server will give the list of files available for visualisation from the directory?

Just specify a directory with a list of rawlog files. That should be sufficient.

>3d animation currently uses trackballcontrols, for changing the pivot the press the right click and move the mouse. There are other controls out there, https://threejs.org/examples/?q=controls#misc_controls_map, but have their own pros and cons.

Cool thanks for clarifying.

> For 3d scan visualisation, are you aware of some rawlog with multiple sensor data. Then I would check how it functions with the RawLogViewer app and modify the app for multiple sources.

It's incredibly useful for a 2.5 D representation from depth cameras or the uncommon case of using multiple lidar scanners at different heights.

Sometimes the map looks different at various heights.
https://www.youtube.com/watch?v=jCFvAOuV_H8

**Comment by rachittibrewal** [comment]
For building mrpt-web, serialization file CSchemeArchiveBase is required from mrpt library. How should I prepare it for merging with MRPT ? This would be required for the video demonstration, or for now I could make the video using mrpt build from my own fork.
These are the works I remember are pending : 
* I am currently facing this [issue](https://github.com/MRPT/mrpt/issues/796) for making the main 2d plots in rawlog viewer app.
* The video is still remaining, I have started work on it.
* Adding extended examples to js library.
* Rawlog loading from directory on server (only a small problem to be solved)

I wanted to finish off work by August 6. I have interviews on campus from 7 - 12 August, it would be very generous if you could help me in doing so. I would like to continue adding more features and working on the project in general after GSoC.

Also GSoC, requires a final submission, in the form of a link. Since this project spans across repositories, can you give an idea of how I should submit the link showing the details of work done. Should I setup a github page?

**Comment by jlblancoc** [comment]
Hi, 
On this: 
> Instructions to launch the RawlogViewer demo app & a link to the most relevant parts of the code?

Did you write some instructions on the ESdoc manual pages? Something really short would be enough... thanks!

On the RawLogViewer issue, I'll check my Eigen version just in case it's newer than yours (now I'm on Windows)...

**Comment by tibrewalrachit** [comment]
Yes will push it by today for launching rawalog app I am making a video.

**Comment by rachittibrewal** [comment]
Yes will push it by today (manual pages) , for launching rawalog app I am making a video.

**Comment by jolting** [comment]
I've recently opened a few of the sample RawLogs in 18.04. You do have to be careful to compile against GTK3 and there are a few tricks you need to get OpenGL to work with wayland(if you use that by default), but besides that no issues.

AFAIK, RawLogViewer doesn't have an issue with the version of Eigen that ships with 18.04.

**Comment by rachittibrewal** [comment]
Update:
* mrpt-web-js [docs](http://mrpt-web.surge.sh) : added examples and brief descriptions in manual
* Created a simple [video](https://www.youtube.com/watch?v=2q5G_dhqSbM&feature=youtu.be) explaining launch and functionality of rawlog-viewer.(Sorry for my poor video editing skills, i will try to make a more fun video later)
* Started work on my final submission [link](https://rachit173.github.io/gsoc2018/).

Any pointers regarding final submission link, like what more verticals should be added, best practices would be great :)

**Comment by jlblancoc** [comment]
Great video!! :+1:

So, apparently you finally solved the crash problem in your rawlog-viewer app, right? It doesn't seem to crash on our tests...

**Comment by rachittibrewal** [comment]
I havent been able to, will try on ubuntu 16.04

**Comment by jolting** [comment]
Good Job! I'm wondering if there is someplace we can host this as a demo.
The javascript part can go on surge.

The websocket server could be containerized.

**Comment by jlblancoc** [comment]
> Good Job! I'm wondering if there is someplace we can host this as a demo.
> The javascript part can go on surge.
>
> The websocket server could be containerized.

Sure, we could host it on mrpt.org in a new subdomain. Just let me
know what do you need exactly, and as far as it's safe we'll do it.

Consider doing it with a front-page which may be the entrypoint to a
number of web demos, even if for now we only have the rawlog-viewer
app...

Cheers!

**Comment by jolting** [comment]
I've cloned all the wiki pages for the following repos.
https://github.com/MRPT/mrpt-web
https://github.com/MRPT/mrpt-web-js
https://github.com/MRPT/rawlog-web-ui

Furthermore, I fixed the unit test for mrpt-web-js. This now uses jest. That was fairly easy to setup.
Also CircleCI now auto deploys to `@mrpt/mrpt-web` package on npmjs.org. That should make using the package a little easier.

**Comment by rachittibrewal** [comment]
Thanks @jolting, I will make the report today and then can continue to work on bugs and issues, after submitting report today.

**Comment by jlblancoc** [comment]
@rachit173 : I just gave you write permissions to all related @MRPT repositories, so you can directly contribute there more easily. You **may** (if you want) use those final URLs in your final report; your contributions are clearly identified there even if that's not your fork, as you know.

A proposal @jolting , @rachit173 : would it make sense to add TravisCI / CircleCI to https://github.com/MRPT/mrpt-web as well?

**Comment by jolting** [comment]
yes

**Comment by jlblancoc** [comment]
I just enabled travis on that repo.

**Comment by rachittibrewal** [comment]
Hi, I am back, thanks for your patience and support. I have submitted report, please have a look at it.
https://rachit173.github.io/gsoc2018/

**Comment by rachittibrewal** [comment]
Thanks for the evaluation and opportunity to work with the organization. I wish to work on the existing project as well as MRPT in general.
